polyply/src/apply_modifications.py

```diff
@@ -12,5 +12,96 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-class ApplyModifications():
-      pass
+from vermouth.log_helpers import StyleAdapter, get_logger
+from vermouth.processors.annotate_mut_mod import parse_residue_spec
+LOGGER = StyleAdapter(get_logger(__name__))
+from .processor import Processor
+import networkx as nx
+
+def _get_protein_termini(meta_molecule):
+    """
+    make a resspec for a protein with correct N-ter and C-ter
+    """
+    max_resid = meta_molecule.max_resid
+    last_node = max_resid - 1
+    last_resname = meta_molecule.nodes[last_node]['resname']
+
+    protein_termini = [({'resid': 1, 'resname': meta_molecule.nodes[0]['resname']}, 'N-ter'),
+                       ({'resid': max_resid, 'resname': last_resname}, 'C-ter')
+                       ]
+
+    return protein_termini
+
+def apply_mod(meta_molecule, modifications):
+    """
+
+    Apply a terminal modification.
+    Note this assumes that the modification is additive, ie. no atoms are
+    removed
+
+    Parameters
+    ----------
+    meta_molecule: :class:`polyply.src.meta_molecule.MetaMolecule`
+    modifications: list
+        list of (resspec, modification) pairs to apply
+
+    Returns
+    ----------
+    meta_molecule
+    """
+
+    molecule = meta_molecule.molecule
+
+    if not molecule.force_field.modifications:
+        LOGGER.warning('No modifications present in forcefield, none will be applied')
+        return meta_molecule
+
+    for target, desired_mod in modifications:
+
+        target_resid = target['resid']
+
+        mod_atoms = {}
+        for mod_atom in molecule.force_field.modifications[desired_mod].atoms:
+            if 'replace' in mod_atom:
+                mod_atoms[mod_atom['atomname']] = mod_atom['replace']
+            else:
+                mod_atoms[mod_atom['atomname']] = {}
+
+        anum_dict = {}
+        for atomid, node in enumerate(molecule.nodes):
```

Comment:
To make this a bit more efficient I would suggest to not loop over every node but directly pick the correct ones. Using the relation that `node = resid - 1` you can do: 

```python
target_residue = meta_molecule.nodes[target_resid -1]
# takes care to skip all residues that come from an itp file
if target_residue.get('from_itp', 'False'):
   continue
# this does not work for choice objects hopefully Peter can give some advise
if target_residue['resname'] != target['resname']:
  continue

# this gives you the correct node indices
for node in target_residue['graph'].nodes:
       aname = molecule.nodes[node]['atomname']
       if aname in mod_atoms.keys():
             anum_dict[aname] = node + 1
             for key, value in mod_atoms[aname].items():
                     molecule.nodes[node][key] = value